core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -336,7 +336,8 @@ class ReplicaManager(val config: KafkaConfig,
                     internalTopicsAllowed: Boolean,
                     isFromClient: Boolean,
                     entriesPerPartition: Map[TopicPartition, MemoryRecords],
-                    responseCallback: Map[TopicPartition, PartitionResponse] => Unit) {
+                    responseCallback: Map[TopicPartition, PartitionResponse] => Unit,
+                    delayedProduceSyncObject: Object = null) {
```

Comment:
Maybe a slightly nicer way to do this is to pass a `DelayedProduce` builder. For example:

`delayedProduceBuilder: (timeout, produceMetadata) -> DelayedProduce`

Then we can remove the `responseCallback` argument since this can be added to the DelayedProduce in the builder.